### PR TITLE
Add local paths to historic layers

### PR DESCRIPTION
**Pull request by SomeoneToIgnore**
Extracted from https://github.com/neondatabase/neon/pull/1603

For scheduling timeline sync tasks, we need to know its local path, since remote_storage knows nothing about pageserver internal kinds of files.

We could go even further and define `HistoricLayer` as

```rust
pub enum HistoricLayer {
    Delta(DeltaLayer),
    Image(ImageLayer),
}
```

the trick to avoid `Arc::clone` and borrow checker issues is to use `HistoricLayer::path` since every such layer is uniquely identified (?) by the local file path.

Alas, both this and the Arc-less versions break test and I'm not really sure why, so any help is appreciated.

### CODE REVIEW

**Comment by SomeoneToIgnore** [comment]
Here's the change removing the `Arc` entirely: https://github.com/neondatabase/neon/compare/kb/historic-layer-local-paths...kb/arcless-historic-layers?expand=1

It does not preserve the layer order during layer removal, but even if I fix that, the tests fail.

@hlinnaka , any proposals on why the current  PR fails the test? Any preferences on how to get the path out of the historic layers? I still can add some `fn path(&self) -> Option<PathBuf>` into `Layer` trait, but not sure if it's a good idea.

**Comment by hlinnaka** [comment]
Hmm, how about just adding `fn path()` to the Layer trait? For in-memory layers, it can just panic or return None or error.

**Comment by hlinnaka** [comment]
> @hlinnaka , any proposals on why the current PR fails the test?

Don't know. But it's the 'wal_removal' test that failed, and there are some errors in the [safekeeper log](https://output.circle-artifacts.com/output/job/d9d0480c-dab8-4d05-acfc-3c2bee9384b1/artifacts/0/tmp/test_output/test_wal_removal/repo/safekeepers/sk1/safekeeper.log).

**Comment by SomeoneToIgnore** [comment]
Ok, now the tests pass here, the culprit seems to be related to metadata reading from disk during compaction loop steps.
Here's the version that spoils the test: https://github.com/neondatabase/neon/compare/kb/historic-layer-local-paths...kb/historic-layer-local-paths-spoiled?expand=1

Not sure why exactly this affects safekeeper and there are no errors in pageserver logs, but feels like those could deadlock involved?

[ARCless PR](https://github.com/neondatabase/neon/compare/kb/historic-layer-local-paths...kb/arcless-historic-layers?expand=1) rebased on top of the fixed version is still failing on benchmarks: https://app.circleci.com/pipelines/github/neondatabase/neon/6079/workflows/35d37a35-6a82-44cc-8f0f-3c0a8f9dc2ff
The error is cryptic, might it be a good thing, considering that Arc + dyn removal improves performance generally?

-----------

Either way, this version is based on Layer trait adjustment and ready for a review.

File: pageserver/src/layered_repository/storage_layer.rs [hunk 1]
```diff
@@ -105,6 +105,9 @@ pub trait Layer: Send + Sync {
     /// log messages, even though they're never not on disk.)
     fn filename(&self) -> PathBuf;
 
+    /// If a layer has a corresponding file on a local filesystem, return its path.
+    fn local_path(&self) -> Option<PathBuf>;
```
[the remark this answers is not in the document]
Comment (SomeoneToIgnore) [reply pageserver/src/layered_repository/storage_layer.rs:109]: Adjusted the comment.

**Review by LizardWizzard — APPROVED** [review]
Looks good to me. Though I'm highly conserned with the spurious test failures

**Pull request merged**